src/platform/nrfconnect/FactoryDataProvider.h

```diff
@@ -0,0 +1,117 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <credentials/DeviceAttestationCredsProvider.h>
+#include <platform/CommissionableDataProvider.h>
+#include <platform/DeviceInstanceInfoProvider.h>
+
+#include <drivers/flash.h>
+#include <pm_config.h>
+
+extern "C" {
+#include "FactoryDataParser.h"
+}
+
+namespace chip {
+namespace DeviceLayer {
+
+struct InternalFlashFactoryData
+{
+    CHIP_ERROR GetFactoryDataPartition(uint8_t *& data, size_t & dataSize)
+    {
+        data     = reinterpret_cast<uint8_t *>(PM_FACTORY_DATA_ADDRESS);
+        dataSize = PM_FACTORY_DATA_SIZE;
+        return CHIP_NO_ERROR;
+    }
+};
+
+struct ExternalFlashFactoryData
+{
+    CHIP_ERROR GetFactoryDataPartition(uint8_t *& data, size_t & dataSize)
+    {
+        int ret = flash_read(mFlashDevice, PM_FACTORY_DATA_ADDRESS, mFactoryDataBuffer, PM_FACTORY_DATA_SIZE);
+
+        if (ret != 0)
+        {
+            return CHIP_ERROR_READ_FAILED;
+        }
+
+        data     = mFactoryDataBuffer;
+        dataSize = PM_FACTORY_DATA_SIZE;
+
+        return CHIP_NO_ERROR;
+    }
+
+    const struct device * mFlashDevice = DEVICE_DT_GET(DT_CHOSEN(zephyr_flash_controller));
+    uint8_t mFactoryDataBuffer[PM_FACTORY_DATA_SIZE];
+};
+
+template <class FlashFactoryData>
+class FactoryDataProvider : public chip::Credentials::DeviceAttestationCredentialsProvider,
+                            public CommissionableDataProvider,
+                            public DeviceInstanceInfoProvider
+{
+public:
+    FactoryDataProvider()
+    {
+        mFactoryData = {};
```

Comment:
I don't think it's needed.

Reply:
Why? Otherwise it will be filled with garbage and in case of invoking getter for some value we might get undefined value. That is especially applicable for some values that are optional, so it might not be present in the factory data, so fields in the structure for those members will not be set after memory partition read.

Reply:
Ah, sorry, I thought it's `mFlashFactoryData`.